source/intel/crc_hw.c

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include <aws/checksums/private/intel/crc32c_compiler_shims.h>
+#include <aws/common/cpuid.h>
+
+static bool detection_performed = false;
+static bool detected_sse42 = false;
+static bool detected_avx512 = false;
+static bool detected_clmul = false;
+
+/*
+ * Computes the Castagnoli CRC32c (iSCSI) of the specified data buffer using the Intel CRC32Q (64-bit quad word) and
+ * PCLMULQDQ machine instructions (if present).
+ * Handles data that isn't 8-byte aligned as well as any trailing data with the CRC32B (byte) instruction.
+ * Pass 0 in the previousCrc32 parameter as an initial value unless continuing to update a running CRC in a subsequent
+ * call.
+ */
+uint32_t aws_checksums_crc32c_hw(const uint8_t *input, int length, uint32_t previousCrc32) {
+
+    if (AWS_UNLIKELY(!detection_performed)) {
+        detected_sse42 = aws_cpu_has_feature(AWS_CPU_FEATURE_SSE_4_2);
+        detected_avx512 = aws_cpu_has_feature(AWS_CPU_FEATURE_AVX512);
+        detected_clmul = aws_cpu_has_feature(AWS_CPU_FEATURE_CLMUL);
+        /* Simply setting the flag true to skip HW detection next time
+           Not using memory barriers since the worst that can
+           happen is a fallback to the non HW accelerated code. */
+        detection_performed = true;
+    }
+
+    uint32_t crc = ~previousCrc32;
+
+    /* For small input, forget about alignment checks - simply compute the CRC32c one byte at a time */
+    if (length < (int)sizeof(slice_ptr_int_type)) {
+        while (length-- > 0) {
+            crc = (uint32_t)_mm_crc32_u8(crc, *input++);
+        }
+        return ~crc;
+    }
+
+    /* Get the 8-byte memory alignment of our input buffer by looking at the least significant 3 bits */
+    int input_alignment = (uintptr_t)(input)&0x7;
+
+    /* Compute the number of unaligned bytes before the first aligned 8-byte chunk (will be in the range 0-7) */
+    int leading = (8 - input_alignment) & 0x7;
+
+    /* reduce the length by the leading unaligned bytes we are about to process */
+    length -= leading;
+
+    /* spin through the leading unaligned input bytes (if any) one-by-one */
+    while (leading-- > 0) {
+        crc = (uint32_t)_mm_crc32_u8(crc, *input++);
+    }
+
+#ifdef AWS_HAVE_AVX512_INTRINSICS
+    int chunk_size = length & ~63;
+
+    if (detected_avx512 && detected_clmul) {
+        if (length >= 256) {
+            crc = aws_checksums_crc32c_avx512(input, length, crc);
+            /* check remaining data */
+            length -= chunk_size;
+            if (!length) {
+                return crc;
+            }
+
+            /* Fall into the default crc32 for the remaining data. */
+            input += chunk_size;
+        }
+    }
+#endif
+
+    if (detected_sse42 && detected_clmul) {
+        return aws_checksums_crc32c_sse42(input, length, crc);
+    }
```

Comment:
Will the sse42 impl be worth invoking after we've finished processing the lead portion using the avx512 implementation?

Reply:
Thank you for the comments. I will review them closely.

For full disclosure - I re-used the avx512 intrinsic implementation from https://chromium.googlesource.com/chromium/src/third_party/zlib/+/b890619bc2b193b8fbe9c1c053f4cd19a9791d92/crc32_simd.c

but recomputed constants for crc32c polynomial :)

Reply:
@pbadari  I've got the build surgery done and tests passing if you'd like to work on the avx512 comments from bdonlan@

Reply:
you'll need the AVX512 branch from aws-c-common until we merge it

Reply:
Yes. I am reviewing the avx512 comments from bdonlan

Reply:
Jonathan, Most of the review comments (from Donlan) for the AVX512 code is further performance improvements which require careful re-write/proto-type and performance analysis. I reached out to our area expert for his input. I am wondering if we can merge the current patch and update it further when we have new code ready? Please let me know.

Reply:
Unfortunately, AVX512 has a history of being an incredibly risky addition to an already functioning (possibly already running really hot) system. Various chipset versions introduce timing issues for side-channel attacks as well as side-effects to other processes sharing the CPU. So we're going to have to run a lot of tests before we can just run this in production anyways, and we'd like the code to be close to structured in the actual final algorithm before running that analysis.

Reply:
Jonathan, I was informed by Arjan (Intel Fellow) that the code submitted is under zlib chromium license which is incompatible with apache-2 license. We are trying to find the best course of action. For now, can we withdraw the patch/submission. I will resolve the issue and re-submit

Reply:
We are submitting AVX-512 based implementation of crc32 to zlib-chromium as well and want to make sure that both code contributions are licensed appropriately. For now, can we withdraw the patch/submission. I will resolve the issue and re-submit.